frontend/components/AddFundsThroughBridge/useGenerateAddFunds.tsx

```diff
@@ -0,0 +1,126 @@
+import { uniq } from 'lodash';
+import { useCallback, useMemo, useState } from 'react';
+
+import { AddressBalanceRecord } from '@/client';
+import { AddressZero } from '@/constants/address';
+import { TokenSymbol } from '@/enums/Token';
+import { useServices } from '@/hooks/useServices';
+import { Address } from '@/types/Address';
+import { BridgeRequest } from '@/types/Bridge';
+import { typedKeys } from '@/types/Util';
+import { getTokenDetailsFromAddress } from '@/utils/middlewareHelpers';
+
+import { DefaultTokenAmount } from './types';
+import { useAddFundsGetBridgeRequirementsParams } from './useAddFundsGetBridgeRequirementsParams';
+
+export type GeneratedInput = {
+  tokenAddress: Address;
+  symbol: TokenSymbol;
+  amount: number;
+};
+
+/**
+ * Generate inputs to add funds for bridging.
+ *
+ * Example: If the chain is Gnosis, user can add OLAS and XDAI.
+ * This function will return an array of objects with tokenAddress, symbol, and amount set to 0.
+ */
+export const useGenerateAddFunds = ({
+  requirements,
+  defaultTokenAmounts,
+  destinationAddress,
+  onlyNativeToken = false,
+}: {
+  requirements: AddressBalanceRecord;
+  defaultTokenAmounts?: DefaultTokenAmount[];
+  destinationAddress?: Address;
+  onlyNativeToken?: boolean;
+}) => {
+  const { selectedAgentConfig } = useServices();
+  const toMiddlewareChain = selectedAgentConfig.middlewareHomeChainId;
+  const getBridgeRequirementsParams =
+    useAddFundsGetBridgeRequirementsParams(destinationAddress);
+
+  // In general, master_safe and master_eoa addresses are used.
+  const allAddresses = typedKeys(requirements);
+
+  // All token addresses that can be used to add funds to the master safe.
+  const tokenAddresses = allAddresses.reduce<Address[]>((acc, address) => {
+    const tokens = requirements[address];
+    if (!tokens) return acc;
+
+    // If onlyNativeToken is true, filter to include only the native token.
+    const typedKeysTokens = typedKeys(tokens).filter((token) =>
+      onlyNativeToken ? token === AddressZero : true,
+    );
+    acc.push(...typedKeysTokens);
+    return acc;
+  }, []);
+
+  const amountsToReceive: GeneratedInput[] = uniq(tokenAddresses).map(
+    (tokenAddress: Address) => {
+      const symbol = getTokenDetailsFromAddress(
+        toMiddlewareChain,
+        tokenAddress,
+      ).symbol;
+
+      // if no default token amounts are provided, set amount to 0.
+      // user can update the amount later.
+      const defaultAmount =
+        defaultTokenAmounts?.find((token) => token.symbol === symbol)?.amount ??
+        0;
+      return { tokenAddress, symbol, amount: defaultAmount };
+    },
+  );
```

Comment:
wonder, why we didn't memoize these, but did memoize bridgeRequirementsParams and inputs related stuff, this will be recalculated anyway when something outside changes or new props are received. and since there are reduces, new array will be created anyway and bridgeRequirementsParams will be triggered again